File: src/main/java/edu/psu/ist/plato/kaiming/arm64/Printer.java
package edu.psu.ist.plato.kaiming.arm64;

import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.Iterator;

import edu.psu.ist.plato.kaiming.BasicBlock;
import edu.psu.ist.plato.kaiming.CFG;
import edu.psu.ist.plato.kaiming.Label;
import edu.psu.ist.plato.kaiming.arm64.LoadStoreInst.AddressingMode;
import edu.psu.ist.plato.kaiming.util.Assert;

public class Printer extends PrintWriter {

    private boolean mIsParseMode;

    public Printer(OutputStream ps, boolean parsemode) {
        super(ps, true);
        mIsParseMode = parsemode;
    }

    public Printer(OutputStream ps) {
        super(ps, true);
        mIsParseMode = false;
    }
    
    private void printSignedHex(long value) {
        if (value < 0) {
            print(String.format("-0x%x", -value));
        } else {
            print(String.format("0x%x", value));
        }
    }

    public void printOpImmediate(Immediate imm) {
        print('#');
        printSignedHex(imm.value());
    }
    
    public void printOpRegister(Register reg) {
        print(reg.id.name());
        if (reg.isShifted()) {
            print(", ");
            Register.Shift shift = reg.shift();
            print(shift.type().name());
            print(", ");
            print(shift.value());
        }
    }
    
    public void printOpMemory(Memory mem) {
        print('[');
        Register base = mem.base();
        if (base != null) {
            printOpRegister(base);
        }
        Memory.Offset off = mem.offset();
        if (off != null) {
            print(", ");
            if (off.isImmediateOffset()) {
                print('#');
                printSignedHex(off.asImmOff().value);
            } else {
                printOpRegister(off.asRegOff().value);
            }
        }
        print(']');
    }
    
    public void printOperand(Operand o) {
        switch (o.type()) {
            case IMMEDIATE:
                printOpImmediate(o.asImmediate());
                break;
            case REGISTER:
                printOpRegister(o.asRegister());
                break;
            case MEMORY:
                printOpMemory(o.asMemory());
                break;
        }
    }
    
    public void printInstruction(Instruction i) {
        if (mIsParseMode) {
            printSignedHex(i.addr());
            print('\t');
        }
        print(i.opcode().rawOpcode());
        print('\t');
        if (i instanceof BranchInst) {
            BranchInst b = (BranchInst)i;
            if (!b.isReturn()) {
                Operand target = b.target();
                if (target.isRegister()) {
                    printOpRegister(target.asRegister());
                } else if (target.isMemory()) {
                    if (target instanceof Relocation) {
                        print(((Relocation)target).targetBlock().label());
                    } else
                        printSignedHex(target.asMemory().offset().asImmOff().value);
                } else {
                    Assert.unreachable();
                }
            }
        } else {
            Iterator<Operand> iter = i.iterator();
            int indexingOperand = -1;
            LoadStoreInst.AddressingMode mode = AddressingMode.REGULAR;
            if (i instanceof LoadStoreInst) {
                mode = ((LoadStoreInst)i).addressingMode();
                if (mode != AddressingMode.REGULAR)
                    indexingOperand = ((LoadStoreInst)i).indexingOperand(); 
            }
            
            int round = 0;
            if (iter.hasNext()) {
                while (true) {
                    if (round == indexingOperand) {
                        switch (mode) {
                            case PRE_INDEX:
                                printOperand(iter.next());
                                print('!');
                                break;
                            case POST_INDEX: {
                                Memory next = iter.next().asMemory();
                                printOperand(next.base());
                                print(", #");
                                printSignedHex(next.offset().asImmOff().value);
                                break;
                            }
                            case REGULAR:
                                Assert.unreachable();
                        }
                    } else 
                        printOperand(iter.next());
                    if (iter.hasNext())
                        print(", ");
                    else
                        break;
                    ++round;
                }
            }
        }
        if (i instanceof CondInstruction) {
            print(", ");
            print(((CondInstruction)i).condition());
        }
    }
    
    public void printLabel(Label label) {
        if (mIsParseMode) {
            printSignedHex(label.addr());
            print('\t');
        }
        print(label.name());
    }
    
    public void printCFG(CFG<Instruction> f) {
        if (mIsParseMode) {
            printLabel((Label)f.entryBlock().label());
            println(':');
        }
        for (BasicBlock<Instruction> bb : f) {
            printBasicBlock(bb);
        }
    }
    
    public void printBasicBlock(BasicBlock<Instruction> bb) {
        if (!mIsParseMode) {
            printLabel(bb.label());
            println(':');
        }
        for (Instruction i : bb) {
            printInstruction(i);
            println();
        }
    }
    
    public void printFunction(Function f) {
        printLabel(f.label());
        println(':');
        for (Instruction i : f.entries()) {
            printInstruction(i);
            println();
        }
    }
}

File: src/main/scala/edu/psu/ist/plato/kaiming/arm64/Printer.scala
package edu.psu.ist.plato.kaiming.arm64

import java.io.PrintWriter
import java.io.OutputStream

import edu.psu.ist.plato.kaiming.Label
import edu.psu.ist.plato.kaiming.BasicBlock
import edu.psu.ist.plato.kaiming.CFG

import edu.psu.ist.plato.kaiming.exception._

class Printer(ps: OutputStream, val parseMode: Boolean) extends PrintWriter(ps, true) {
  
  def this(ps: OutputStream) = this(ps, false)
  
  private def printSignedHex(value: Long) {
    if (value < 0)
      print("-0x%x".format(-value))
    else
      print("0x%x".format(value))
  }
  
  def printOpImmediate(imm: Immediate) {
    print('#')
    printSignedHex(imm.value)
  }
  
  def printOpRegister(reg: Register) {
    print(reg.id.entryName)
    reg.shift match {
      case Some(shift) =>
        print(", ")
        print(shift.ty.entryName)
        print(", ")
        print(shift.value)
      case None => 
    }
  }
  
  def printOpMemory(mem: Memory) {
    print('[')
    mem.base match {
      case Some(base) => printOpRegister(base)
      case None =>
    }
    mem.off match {
      case Some(off) =>
        print(", ")
        off match {
          case Left(imm) => printOpImmediate(imm)
          case Right(reg) => printOpRegister(reg)
        }
      case None =>
    }
    print(']')
  }
  
  def printOperand(o: Operand) = {
    o match {
      case r: Register => printOpRegister(r)
      case i: Immediate => printOpImmediate(i)
      case m: Memory => printOpMemory(m)
    }
  }
  
  def printInstruction(i: Instruction) {
    if (parseMode) {
      printSignedHex(i.addr)
      print('\t')
    }
    print(i.opcode.rawcode)
    print('\t')
    i match {
      case b: BranchInst =>
        if (!b.isReturn) {
          b.relocatedTarget match {
            case Some(b) => printLabel(b.label)
            case None => 
              b.target match {
                case r: Register => printOpRegister(r)
                case Memory(base, Some(Left(off))) => 
                  printSignedHex(off.value)
                case _ => throw new UnreachableCodeException()
            }
          }
        }
      case _ => {
        import Instruction.AddressingMode._
        val (indexingOperand, addressingMode) = 
          if (i.isInstanceOf[LoadStoreInst]) {
            val lsi = i.asInstanceOf[LoadStoreInst]
            val m = lsi.addressingMode
            val idx = lsi.indexingOperand
            (if (m == Regular) -1 else idx, m)
          } else {
            (-1, Regular)
          }
        (0 until i.operands.size).foreach {
          n => 
            if (n == indexingOperand) {
              addressingMode match {
                case PreIndex =>
                  printOperand(i.operands(n))
                  print('!')
                case PostIndex => {
                  val next = i.operands(n).asMemory
                  printOperand(next.base.get)
                  print(", ")
                  printOpImmediate(next.off.get.left.get)
                }
                case Regular =>
                  throw new UnreachableCodeException()
              }
            } else {
              printOperand(i.operands(n))
            }
            if (n != i.operands.size - 1)
              print(", ")
        }
      }
    }
    if (i.isInstanceOf[SelectInst]) {
      print(", ")
      print(i.asInstanceOf[SelectInst].condition.entryName)
    }
  }
  
  def printLabel(l: Label) {
    if (parseMode) {
      printSignedHex(l.addr)
      print('\t')
    }
    print(l.name)
  }
  
  def printBasicBlock(bb: BasicBlock[Instruction]) {
    if (!parseMode) {
      printLabel(bb.label)
      println(':')
    }
    bb.foreach { i => printInstruction(i); println() }
  }
  
  def printCFG(cfg: CFG[Instruction]) {
    printLabel(cfg.parent.label)
    println(':');
    cfg.foreach { bb => printBasicBlock(bb) }
  }
  
  def printFunction(f: Function) {
    printLabel(f.label)
    println(':')
    f.entries.foreach { i => printInstruction(i); println() }
  }
  
}